### ISSUE TITLE
Reimplement Smart route recalculation if possible

### ISSUE THREAD

**Issue by Guylby** [osmandapp/OsmAnd#19737]
### Describe the idea (required)

Hello,
First, thanks for all your work on the new offline routing, which is for sure more convenient and faster than the old one.

However, for long trips or trips that use certain routing parameters (Fuel Efficient Way, No Toll Roads for example), routing can still be counted at more than 20-30s.

This is acceptable during the initial calculation, but if a reroute happens at the start of the trip (road closed, taking a detour or taking the wrong road), you will have to recalculate the whole trip again. 
In an unfamiliar city with many directions and intersections close to one another, the recalculation can finish after you have already taken a wrong turn, leading to another recalculation etc...
Smart route recalculation could help in these scenarios and with the performance of the new HH routing be near instantaneous!

It would be the icing on the cake and really help performance!

### Tell us about the expected behaviour (required)

Reimplement Smart route recalculation

### Tell us about alternatives you've considered (required)

No real alternative

**Comment by vshcherb** [comment]
Yes Smart recalculation was initially implemented for first 20-30 km, later we used by default only after 2 intermediate destination. The issue with "smart" it wasn't generated correct optimal routes, we think to do something in the future once we identify most typical use cases

**Comment by Guylby** [comment]
For my use in Car routing, this is where this feature would be most useful:

- Taking a small detour: during a long trip you often take small detours without changing the main route, for example: 
Taking a highway exit for refueling/taking a break,
Stopping in a town for food/break/etc...

- Roadworks/roads closed that force you to take an alternate route.

- Incorrect navigation: sometimes you just miss the turn in a town, arrive in a confusion intersection, or you are stuck in traffic and can't change lane.

All are small detours that are less than 10kms but causes a lot of route recalculation for the entire itinerary, when you just want to get back to the main itinerary as soon as possible.
